### ISSUE TITLE
Alternative mode to handle requests

### ISSUE THREAD

**Issue by mingodad** [h2o/h2o#142]
Hello !

Trying to use h2o as an application server I found that the actual implementation using threads to manage requests without a way to put a connection on a kind of wait state to process other requests on the same thread a limiting factor.

Ideally I would expect to be able to manage multiple connections per thread to reduce latency and achieve higher throughput.

The description here is the mix of multiple threads that can manage multiple requests using queues based on the socket events. I'm trying to layout the idea and hope to evolve it and appreciate any feedback/suggestions on it.  

I'm not sure the best way to do it:
- One possible way would be to have the listeners accept been processed by only one thread that would dispatch the accepted sockets to queues and the threads would peak then to process, if they need to write a response that would block (because it's size is greater than the network buffer size) they could put that socket on another queue till it get ready to write again and meanwhile process another request. Also if to satisfy one request it needs to make another request externa/internal (databases, call external apis, memcached, redis, ...) they would put that request on another queue and start processing another request.
- Other way is to somehow have on the h2o library a generic socket queue that can be used by the applications to associate a callback processor that would be called when the enqeued sockets become ready for read/write, for example when processing a proxy call we could start a queue to wait for the response form the proxied server and if we have more calls to the same request insert on that queue to prevent multiple identical calls and start processing another requests.

**Comment by kazuho** [comment]
@mingodad 
> Trying to use h2o as an application server I found that the actual implementation using threads to manage requests without a way to put a connection on a kind of wait state to process other requests on the same thread a limiting factor.

There is no such restriction.  The guideline for writing a generator is:

* return `0` from `h2o_handler_t::on_req` to capture the request
* call `h2o_start_response` when the response header becomes ready (at the same time register the generator)
* perform appropriate actions when `h2o_generator_t::proceed` or `h2o_generator_t::stop` are called

The three can be performed asynchronously.

> Other way is to somehow have on the h2o library a generic socket queue that can be used by the applications to associate a callback processor that would be called when the enqeued sockets become ready for read/write, for example when processing a proxy call we could start a queue to wait for the response from the proxied server and if we have more calls to the same request insert on that queue to prevent multiple identical calls and continue processing another requests.

Yes.  This is how the reverse proxy (lib/handler/proxy.c) is implemented.  I would suggest reading it.

**Comment by mingodad** [comment]
It seems that it's what I'm looking for but can not see it clear, could you give an example for these two scenarios:

Pseudo code it's ok although a complete example would even be better ! 

- 1 we have a post with a potential big body:
   - I want to be notified at the start of request to specify the body size limit based on parameters passed, identify user and check access rights.
  - Then if I found that is not ok to process this request return a response from here explaining why.
  - Ok I accept the request and because it's body is big I'll want to enqueue it to process other requests meanwhile.

2- I have a request with a big payload, so I do not want to be blocked while writing to it, how I enqueue it and start managing other requests in parallel.

**Comment by mingodad** [comment]
Also do the proxy coalesce the requests ? Or if it receives x simultaneos requests to the same resource it makes x calls ?

Also would be nice to have a configuration example on how to use/configure the proxy, actually there isn't any so will be very unlikely that anyone will use it, unless based on try/error reverse engineer the proxy.c and proxy configurator.

Cheers !

**Comment by kazuho** [comment]
> I want to be notified at the start of request to specify the body size limit based on parameters passed, identify user and check access rights.

That is not possible at the moment.  Handlers are called by the protocol handlers _after_ all the request body has been received.

> Then if I found that is not ok to process this request return a response from here explaining why.

Once if the previous issue gets resolved, it would be possible.  However you should not expect that clients in general will be able to render such responses.  Many HTTP/1 clients do not read the response until they succeed in sending the request entity.  In HTTP/2, such behavior becomes impossible as the specficitaion suggests the server implementations to send `RST_STREAM` frame (indicating the client that the stream has been closed) in such case.

> Ok I accept the request and because it's body is big I'll want to enqueue it to process other requests meanwhile.

The request body is enqueued enqueued by the protocol handlers of H2O, and is available as `h2o_req_t::entity` to the handlers.

> I have a request with a big payload, so I do not want to be blocked while writing to it, how I enqueue it and start managing other requests in parallel.

You do not need to take care of the issue.  `h2o_send` accepts an arbitrary length of data and sends them eventually.  After all the data has been sent, it calls `h2o_generator_t::proceed` to request additional data.  When the function gets called, the genenator should call `h2o_send` immediately (or asynchronously) to feed more data to the protocol handlers (as well as setting the `is_final` flag to indicate that all data has been emitted).

> Also do the proxy coalesce the requests ?

No; for the time being.

> Also would be nice to have a configuration example on how to use/configure the proxy, actually there isn't any, so will be very unlikely that anyone will use it, unless based on try/error reverse engineer the proxy.c and proxy configurator.

Thank you for the suggestion.  It is clear that docs should be written.
